### PR TITLE
feat: Interest and Reserves

### LINKED ISSUE #98: x/leverage: Create Interest Rate Algo
## Summary

Create the initial interest rate algorithm for `x/leverage` that's based on the supply of the token.

----

## For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### LINKED ISSUE #144: x/leverage: Reconcile Params
## Summary

Currently, the `x/leverage` module has the following parameters per token in its registry:

- base_denom
- exchange_rate
- collateral_weight
- base_borrow_rate

Compound has the following per token in their governance model:

- Collateral Factor, e.g. 0%
- Reserve Factor, e.g. 25%
- COMP Speed, e.g. 0.0014625
- APY borrow base rate, e.g. 2% 
- APY borrow rate at kink, e.g. 20%
- Kink at 80% utilization
- 100% APY borrow rate at 100% utilization

We need to reconcile what parameters we're missing and if we need them. Based on the above list, they seem to be:

- Reserve Factor, e.g. 25%
- COMP Speed, e.g. 0.0014625
- kink and utilization related fields

----

## For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by toteki**
## Description

Implement dynamic-rate interest on borrows, reserves that increase when borrow interest accrues, and a token:uToken exchange rate that converts the remaining (i.e. unreserved) accrued interest into lender rewards.

Closes #98 
Closes #144 

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/genesis.go [hunk 1]
```diff
@@ -11,7 +11,6 @@ import (
 // state.
 func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
 	k.SetParams(ctx, genState.Params)
```
Comment (toteki) [inline x/leverage/genesis.go:13]:
```suggestion
	k.SetParams(ctx, genState.Params)
	
```

File: x/leverage/keeper/exchangerate.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	key := types.CreateExchangeRateKey(denom)
+	amount := sdk.ZeroDec()
+	bz := store.Get(key)
+	if bz == nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrInvalidAsset, denom)
+	}
+	err := amount.Unmarshal(bz)
+	if err != nil {
+		panic(err)
+	}
```
Comment (alexanderbez) [inline x/leverage/keeper/exchangerate.go:50]: Why does this return an error and panic?

File: x/leverage/keeper/exchangerate.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+func (k Keeper) SetExchangeRate(ctx sdk.Context, denom string, rate sdk.Dec) error {
+	store := ctx.KVStore(k.storeKey)
+	if !k.IsAcceptedToken(ctx, denom) {
+		return sdkerrors.Wrap(types.ErrInvalidAsset, denom)
+	}
+	rateKey := types.CreateExchangeRateKey(denom)
+	bz, err := rate.Marshal()
+	if err != nil {
+		panic(err)
+	}
```
Comment (alexanderbez) [inline x/leverage/keeper/exchangerate.go:64]: ditto on error + panic.

File: x/leverage/types/keys.go [hunk 4]
```diff
@@ -88,3 +99,37 @@ func CreateCollateralSettingKey(borrowerAddr sdk.AccAddress, uTokenDenom string)
...
+// CreateExchangeRateKey returns a KVStore key for getting and setting the token:uToken rate for a a given token.
+func CreateExchangeRateKey(tokenDenom string) []byte {
+	// exchangeRatePrefix | denom | 0x00
+	var key []byte
+	key = append(key, KeyPrefixExchangeRate...)
+	key = append(key, []byte(tokenDenom)...)
+	return append(key, 0) // append 0 for null-termination
+}
+
+// CreateExchangeRateKeyNoDenom returns a safe copy of exchangeRatePrefix
```
Comment (alexanderbez) [inline x/leverage/types/keys.go:129]: Why is this needed? Can't you just call and directly use `KeyPrefixExchangeRate`

File: x/leverage/keeper/exchangerate.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	store := ctx.KVStore(k.storeKey)
+	totalBorrows, err := k.GetTotalBorrows(ctx)
+	if err != nil {
+		return err
+	}
+	// Calculate exchange rates for all denoms which have an exchange rate stored in keeper.
+	// If a token is registered but its exchange rate is never initialized (set to 1.0), this
+	// iterator will fail to detect it, and its exchange rate will remain undefined.
+	exchangeRatePrefix := types.CreateExchangeRateKeyNoDenom()
+	iter := sdk.KVStorePrefixIterator(store, exchangeRatePrefix)
```
Comment (alexanderbez) [inline x/leverage/keeper/exchangerate.go:80]: Why not just use `KeyPrefixExchangeRate` directly?

File: x/leverage/keeper/exchangerate.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
+package keeper
```
Comment (alexanderbez) [inline x/leverage/keeper/exchangerate.go:1]: Can we call this file `exchange_rate.go`

File: x/leverage/keeper/keeper.go [hunk 7]
```diff
@@ -95,115 +96,58 @@ func (k Keeper) WithdrawAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, uToken
...
+
+	withdrawal, err := k.ExchangeUTokens(ctx, uToken)
+	if err != nil {
+		return err
+	}
+
+	// Ensure module account has sufficient unreserved tokens to withdraw
+	reservedAmount := k.GetReserveAmount(ctx, withdrawal.Denom)
+	availableAmount := k.bankKeeper.GetBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), withdrawal.Denom).Amount
+	if withdrawal.Amount.GT(availableAmount.Sub(reservedAmount)) {
```
Comment (alexanderbez) [inline x/leverage/keeper/keeper.go:110]: Do we need to worry about negative values here?

File: x/leverage/keeper/math.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package keeper
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
+// Interpolate takes a line defined by two points (xMin, yMin) and (xMax, yMax), then finds the y-value of the
+// point on that line for an input x-value. It will return yMin if xMin = xMax (i.e. a vertical line).
+// While this function is intended for interpolation (xMin < x < xMax), it works correctly even when x is outside
+// that range or when xMin > xMax.
+func (k Keeper) Interpolate(ctx sdk.Context, x, xMin, yMin, xMax, yMax sdk.Dec) sdk.Dec {
```
Comment (alexanderbez) [inline x/leverage/keeper/math.go:9]: Do we have a unit test for this function?

File: x/leverage/keeper/reserves.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+		err := amount.Unmarshal(bz)
+		if err != nil {
+			panic(err)
+		}
+	}
+	return amount
+}
+
+// IncreaseReserves adds an sdk.Coin (denom, amount) to the module's reserve requirements.
+func (k Keeper) IncreaseReserves(ctx sdk.Context, coin sdk.Coin) error {
```
Comment (alexanderbez) [inline x/leverage/keeper/reserves.go:25]: ditto on panic/error combo

File: x/leverage/types/genesis.go [hunk 10]
```diff
@@ -1,16 +1,25 @@
...
+			InterestEpoch: 100,
+		},
 		Registry: []Token{},
 	}
 }
 
 // Validate performs basic genesis state validation returning an error upon any
 // failure.
 func (gs GenesisState) Validate() error {
+
```
Comment (alexanderbez) [inline x/leverage/types/genesis.go:18]:
```suggestion
```

File: x/leverage/types/genesis.go [hunk 11]
```diff
@@ -1,16 +1,25 @@
...
+	if gs.Params.InterestEpoch <= 0 {
+		return fmt.Errorf("interest epoch must be positive: %d", gs.Params.InterestEpoch)
+	}
```
Comment (alexanderbez) [inline x/leverage/types/genesis.go:21]: We should instead call `params.Validate()`

File: x/leverage/types/genesis.go [hunk 12]
```diff
@@ -1,16 +1,25 @@
 package types
 
+import fmt "fmt"
+
 // DefaultGenesis returns the default genesis state of the x/leverage module.
 func DefaultGenesis() *GenesisState {
 	return &GenesisState{
-		Params:   Params{},
+		Params: Params{
+			InterestEpoch: 100,
```
Comment (alexanderbez) [inline x/leverage/types/genesis.go:9]:
```suggestion
			InterestEpoch: DefaultInterestEpoch,
```

File: x/leverage/keeper/exchangerate.go (hunk 2 above)
Comment (toteki) [reply x/leverage/keeper/exchangerate.go]: Fixed, along with the rest of the panic(err) instances in error-returning functions in `x/leverage`

File: x/leverage/types/keys.go (hunk 4 above)
Comment (toteki) [reply x/leverage/types/keys.go:129]: Note: As discussed on slack, returning a copy with a new underlying array for slice safety.

File: x/leverage/keeper/keeper.go [hunk 13]
```diff
@@ -72,14 +71,16 @@ func (k Keeper) LendAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, loan sdk.C
...
 	// mint uTokens
-	// TODO: Use exchange rate instead of 1:1 redeeming
-	uToken := sdk.NewCoin(k.FromTokenToUTokenDenom(ctx, loan.Denom), loan.Amount)
+	uToken, err := k.ExchangeTokens(ctx, loan)
+	if err != nil {
+		return err
+	}
 	uTokens := sdk.NewCoins(uToken)
-	if err := k.bankKeeper.MintCoins(ctx, types.ModuleName, uTokens); err != nil {
+	if err = k.bankKeeper.MintCoins(ctx, types.ModuleName, uTokens); err != nil {
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/keeper/keeper.go:85]: Both ways seem to work, though I do see more `:=` in the wild.

While I can't find any reasons, I'll go ahead and change back to shadowing here just to be more standard.

File: x/leverage/keeper/keeper.go (hunk 7 above)
Comment (toteki) [reply x/leverage/keeper/keeper.go:116]: BankKeeper.GetBalance should be fine, but I'll put a guard on GetReserveAmount to make sure it never returns a negative value. 

(In practice it should be fine, as reserves only have `Get` and `Increase` functions, so achieving a negative value would have to occur by int overflow or a marshaling problem)

File: x/leverage/keeper/keeper.go [hunk 14]
```diff
@@ -95,115 +96,58 @@ func (k Keeper) WithdrawAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, uToken
...
 	}
 
-	// ensure the tokens exist in the leverage module account's balance
-	// TODO: Use exchange rate instead of 1:1 redeeming
-	tokenDenom := k.FromUTokenToTokenDenom(ctx, uToken.Denom)
-	token := sdk.NewCoin(tokenDenom, uToken.Amount)
-	if !k.bankKeeper.HasBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), token) {
-		// TODO: We should never enter this case -- consider a panic
-		return sdkerrors.Wrap(types.ErrInsufficientBalance, token.String())
+	// TODO: Calculate lender's borrow limit and borrowed value, if any, to prevent
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/keeper/keeper.go]: Created issue #213 for this and two other TODOs in the code, which involve using borrow limit to check transactions.
Such comments now start as `// TODO #213:`

File: x/leverage/keeper/math.go (hunk 8 above)
Comment (toteki) [reply x/leverage/keeper/math.go]: Yes - it was in `keeper_test.go`, but I just separated it out into a new file `math_test.go`

File: x/leverage/types/genesis.go (hunk 11 above)
Comment (toteki) [reply x/leverage/types/genesis.go]: (done)

File: x/leverage/keeper/interest.go [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+	prevInterestTime := sdk.ZeroInt()
+	bz := store.Get(timeKey)
+	err := prevInterestTime.Unmarshal(bz)
+	if err != nil {
+		panic(err)
+	}
+
+	// Calculate time elapsed since last interest accrual (measured in years for APR math)
+	currentTime := ctx.BlockTime().Unix()
+	yearsElapsed := sdk.NewDec(currentTime - prevInterestTime.Int64()).QuoInt64(31536000)
```
[the remark this answers is not in the document]
Comment (toteki) [reply x/leverage/keeper/interest.go]: added to `x/leverage/types`

**Review by justyntemme — APPROVED** [review]

File: x/leverage/keeper/math_test.go [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestInterpolate(t *testing.T) {
+
```
Comment (alexanderbez) [inline x/leverage/keeper/math_test.go:11]:
```suggestion
```

File: x/leverage/keeper/reserves.go [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/umee-network/umee/x/leverage/types"
+)
+
+// GetReserveAmount gets the amount reserved of a specified token. On invalid asset, the reserved amount is zero.
+func (k Keeper) GetReserveAmount(ctx sdk.Context, denom string) sdk.Int {
+	store := ctx.KVStore(k.storeKey)
+	key := types.CreateReserveAmountKey(denom)
+	amount := sdk.ZeroInt()
+	bz := store.Get(key)
+	if bz != nil {
```
Comment (alexanderbez) [inline x/leverage/keeper/reserves.go:15]: Curious, can we bubble up the error or do you think it's wise to panic here?

**Review by alexanderbez — APPROVED** [review]

File: x/leverage/keeper/reserves.go (hunk 17 above)
Comment (toteki) [reply x/leverage/keeper/reserves.go:15]: The panics here are on two cases:
- negative reserve amount
- failed to unmarshal sdk.Int (but it wasn't empty/nil)

Since those two cases shouldn't occur at all and the rest of the function works without an error return at all, I think we can keep it as a panic. This resembles how `bankkeeper.GetBalance` works - no error return, potential panic on unmarshal fail.

Comment (alexanderbez) [reply x/leverage/keeper/reserves.go:15]: ACK
